packages/builders/src/components/button/Button.ts

```diff
@@ -1,151 +1,34 @@
-	emojiValidator,
-	urlValidator,
-	validateRequiredButtonParameters,
-} from '../Assertions.js';
+import type { APIButtonComponent } from 'discord-api-types/v10';
+import { isValidationEnabled } from '../../util/validation.js';
+import { buttonPredicate } from '../Assertions.js';
 import { ComponentBuilder } from '../Component.js';
 
 /**
  * A builder that creates API-compatible JSON data for buttons.
  */
-export class ButtonBuilder extends ComponentBuilder<APIButtonComponent> {
-	/**
-	 * Creates a new button from API data.
-	 *
-	 * @param data - The API data to create this button with
-	 * @example
-	 * Creating a button from an API data object:
-	 * ```ts
-	 * const button = new ButtonBuilder({
-	 * 	custom_id: 'a cool button',
-	 * 	style: ButtonStyle.Primary,
-	 * 	label: 'Click Me',
-	 * 	emoji: {
-	 * 		name: 'smile',
-	 * 		id: '123456789012345678',
-	 * 	},
-	 * });
-	 * ```
-	 * @example
-	 * Creating a button using setters and API data:
-	 * ```ts
-	 * const button = new ButtonBuilder({
-	 * 	style: ButtonStyle.Secondary,
-	 * 	label: 'Click Me',
-	 * })
-	 * 	.setEmoji({ name: '🙂' })
-	 * 	.setCustomId('another cool button');
-	 * ```
-	 */
-	public constructor(data?: Partial<APIButtonComponent>) {
-		super({ type: ComponentType.Button, ...data });
-	}
-
-	/**
-	 * Sets the style of this button.
-	 *
-	 * @param style - The style to use
-	 */
-	public setStyle(style: ButtonStyle) {
-		this.data.style = buttonStyleValidator.parse(style);
-		return this;
-	}
-
-	/**
-	 * Sets the URL for this button.
-	 *
-	 * @remarks
-	 * This method is only available to buttons using the `Link` button style.
-	 * Only three types of URL schemes are currently supported: `https://`, `http://`, and `discord://`.
-	 * @param url - The URL to use
-	 */
-	public setURL(url: string) {
-		(this.data as APIButtonComponentWithURL).url = urlValidator.parse(url);
-		return this;
-	}
-
-	/**
-	 * Sets the custom id for this button.
-	 *
-	 * @remarks
-	 * This method is only applicable to buttons that are not using the `Link` button style.
-	 * @param customId - The custom id to use
-	 */
-	public setCustomId(customId: string) {
-		(this.data as APIButtonComponentWithCustomId).custom_id = customIdValidator.parse(customId);
-		return this;
-	}
-
-	/**
-	 * Sets the SKU id that represents a purchasable SKU for this button.
-	 *
-	 * @remarks Only available when using premium-style buttons.
-	 * @param skuId - The SKU id to use
-	 */
-	public setSKUId(skuId: Snowflake) {
-		(this.data as APIButtonComponentWithSKUId).sku_id = skuId;
-		return this;
-	}
-
-	/**
-	 * Sets the emoji to display on this button.
-	 *
-	 * @param emoji - The emoji to use
-	 */
-	public setEmoji(emoji: APIMessageComponentEmoji) {
-		(this.data as Exclude<APIButtonComponent, APIButtonComponentWithSKUId>).emoji = emojiValidator.parse(emoji);
-		return this;
-	}
+export abstract class ButtonBuilder<ButtonData extends APIButtonComponent> extends ComponentBuilder<ButtonData> {
```

Comment:
Style is the same property as any other. There should be a way to dynamically change it

Reply:
No, it's not. Link style and Premium style differ froim the other styles in terms of what properties they need or accept.

Reply:
I found you previously discussed this. Yeah, `new CustomIdButton().setStyle(ButtonStyle.Primary)` feels weird, but nobody says you should remove the style-based classes. Regarding `ContextMenuCommandBuilder.setType()`, you don't usually change the type, but the style is not the type. Sometimes, you need to change the style. For example, a switch button that switches between `Success` and `Danger`, meaning it's on or off. Creating a new component instead of using `.setStyle() feels more weird.

Reply:
> No, it's not. Link style and Premium style differ from the other styles in terms of what properties they need or accept.

The accepting style can be validated

Reply:
As an alternative there can be `CustomIdButtonBuilder.from()` within `/builders` to change the style of the builders

Reply:
> Creating a new component instead of using `.setStyle()` feels more weird.

Debatable. To me it seems perfectly reasonable considering different button styles may have different properties

> The accepting style can be validated

We are only doing validation on `toJSON()` calls

I don't see a compelling reason for why we should specifically create a less type-safe (and runtime-safe for that matter) Button builder if everything else does not

> As an alternative there can be `CustomIdButtonBuilder.from()` within `/builders` to change the style of the builders

I believe that would be opening up a can of worms for the rest of the package. If we're adding `CustomIdButtonBuilder.from()`, why shouldn't we add `Button.from()` and `SelectMenu.from()`. Oh and then let's also add `Component.from()`

The fact that `CustomIdButtonBuilder` currently is in fact a builder and not a mixin is simply an implementation detail. It *could* be made to be just a mixin (should probably be done for consistency reasons with the rest of the package)

Reply:
At this point this pull request has literally exploded. Please, let's just get this merged as-is and we can iterate depending on how it feels to use, otherwise I'm never gonna get to work on the needed follow-ups (mainlib impl and such).

Reply:
Ok, I'm fine with this. This can be resolved